content/assignments/memory-allocator/tests/checker.py

```diff
@@ -0,0 +1,263 @@
+        Call.__init__(self, line)
+        self.syscalls = syscalls.copy()
+
+    def __repr__(self) -> str:
+        ret = super().__repr__()
+        if self.syscalls:
+            ret += '\n  ' + '\n  '.join([str(s) for s in self.syscalls])
+        return ret
+
+
+def prettyMmapArgs(mmapSyscall: SysCall):
+    mmapSyscall.args[2] = ' | '.join(list(map(lambda x: x[0], filter(lambda prot: int(mmapSyscall.args[2]) & prot[1], SysCall.MMAP_PROTS.items()))))
+    mmapSyscall.args[3] = ' | '.join(list(map(lambda x: x[0], filter(lambda flag: int(mmapSyscall.args[3]) & flag[1], SysCall.MMAP_FLAGS.items()))))
+
+
+def parseLtraceOutput(ltraceOutput: str):
+    # Filter lines that do not contain traced calls
+    lines = list(filter(lambda line: any([line.find(tc) != -1 for tc in TRACED_CALLS]), ltraceOutput.splitlines()))
+    tracedCalls = []
+    unfinishedStack = []
+    syscallsBatch = []
+
+    # Get heap start
+    heapStart = int(SysCall(lines[0]).ret, 16)
+    # Get exit status
+    exitStatus = ltraceOutput.splitlines()[-1]
+
+    # Store addresses
+    heapAddresses = {}
+    mappedAddresses = {}
+
+    # Extract libcalls and nested syscalls from output
+    for line in lines:
+        # Syscalls
+        if line.startswith('SYS_'):
+            # Ignore syscalls made outside a library call
+            if unfinishedStack:
+                syscallsBatch.append(SysCall(line))
+                # Update mmap params
+                if syscallsBatch[-1].name == 'mmap':
+                    prettyMmapArgs(syscallsBatch[-1])
+            continue
+
+        # Libcalls
+        if '<unfinished' in line:
+            unfinishedStack.append(line)
+        elif 'resumed>' in line:
+            if not unfinishedStack:
+                print('No call to resume')
+                exit(-1)
+
+            newLine = unfinishedStack.pop()
+            newLine = newLine[:newLine.find('<')] + line[line.find('>')+1:]
+
+            # Only add top level calls
+            if not unfinishedStack:
+                tracedCalls.append(LibCall(newLine, syscallsBatch))
+                syscallsBatch.clear()
+        elif not unfinishedStack:
+            # Add top level libcalls
+            tracedCalls.append(LibCall(line))
+
+    # Map addresses to relative values
+    for libcall in tracedCalls:
+        # Syscalls
+        for syscall in libcall.syscalls:
+            # Mapped addresses
+            if syscall.name == 'mmap' and syscall.ret not in mappedAddresses:
+                index = 1 + len(list(filter(lambda v: '+' not in v, [v for v in mappedAddresses.values()])))
+                mappedAddresses[syscall.ret] = f'<mapped-addr{index}>'
+            # Heap addresses
+            elif syscall.name == 'brk' and syscall.ret not in heapAddresses:
+                heapAddresses[syscall.ret] = 'HeapStart + ' + hex(int(syscall.ret, 16) - heapStart)
+
+        # Return values
+        if libcall.ret != '<void>' and libcall.ret != '0':
```

Comment:
Seems that `'nil'` is also a not valid ret value.
```suggestion
        if libcall.ret != '<void>' and libcall.ret != '0' and libcall.ret != 'nil':
```